sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -483,41 +532,115 @@ public void validate(PipelineOptions options) {
+                new DoFn<String, String>() {
+                  @ProcessElement
+                  public void processElement(ProcessContext c)
+                      throws Exception {
+                    String jobUuid = c.element();
+                    BigQuerySourceBase source = applySourceTransform(
+                        jobUuid);
+                    String schema = BigQueryHelpers.toJsonString(
+                        source.getSchema(c.getPipelineOptions()));
+                    c.output(tableSchemaTag, schema);
+                    List<ResourceId> files = source.extractFiles(c.getPipelineOptions());
+                    for (ResourceId file : files) {
+                      c.output(file.toString());
+                    }
+                  }
+                })
+            .withOutputTags(filesTag, TupleTagList.of(tableSchemaTag)));
+        final PCollectionView<String> schemaView = tuple.get(tableSchemaTag)
+            .apply(View.<String>asSingleton());
+        rows = tuple.get(filesTag)
+            .apply(WithKeys.of(new SerializableFunction<String, String>() {
+                public String apply(String s) {
+                  return s;
+                }
+              }))
+            .apply(Reshuffle.<String, String>of())
+            .apply(Values.<String>create())
+            .apply("ReadFiles", ParDo.of(
+                new DoFn<String, TableRow>() {
+                  @ProcessElement
+                  public void processElement(ProcessContext c)
+                      throws Exception {
+                    TableSchema schema = BigQueryHelpers.fromJsonString(
+                        c.sideInput(schemaView), TableSchema.class);
+                    String jobUuid = (String) c.sideInput(jobIdTokenView);
+                    BigQuerySourceBase source = applySourceTransform(jobUuid);
+                    List<BoundedSource<TableRow>> sources =
+                    source.createSources(ImmutableList.of(
+                        FileSystems.matchNewResource(
+                            c.element(), false /* is directory */)), schema);
+                    for (BoundedSource<TableRow> avroSource : sources) {
+                      BoundedSource.BoundedReader<TableRow> reader =
+                          avroSource.createReader(c.getPipelineOptions());
+                      if (reader.start()) {
+                        c.output(reader.getCurrent());
+                        while (reader.advance()) {
+                          c.output(reader.getCurrent());
+                        }
+                      }
+                    }
+                  }
+                }).withSideInputs(schemaView, jobIdTokenView));
       }
       PassThroughThenCleanup.CleanupOperation cleanupOperation =
           new PassThroughThenCleanup.CleanupOperation() {
             @Override
-            void cleanup(PipelineOptions options) throws Exception {
+            void cleanup(DoFn.ProcessContext c) throws Exception {
```

Comment:
Not sure I'm getting this right -- if I want to remove the sideInput cast, I would type that call, but that seems orthogonal to the DoFn<> typing.